File: examples/backend_registry_demo.cpp
/**
 * @file backend_registry_demo.cpp
 * @brief Simple demo of the BackendRegistry system
 *
 * Shows how to create backends at runtime without hardcoded type checking.
 */

#include <iostream>
#include <memory>

#include "trossen_sdk/trossen_sdk.hpp"

int main() {
  // List registered backends
  std::cout << "Registered backends:\n";
  for (const auto& type : {"null", "mcap", "lerobot", "trossen"}) {
    if (trossen::io::BackendRegistry::is_registered(type)) {
      std::cout << "  Is Registered: " << type << "\n";
    }
  }
  std::cout << "\n";

  // Create and use a backend through the registry
  std::cout << "Creating null backend...\n";
  trossen::io::backends::NullBackend::Config cfg;
  cfg.type = "null";

  auto backend = trossen::io::BackendRegistry::create("null", cfg);
  backend->open();

  // Write some data
  trossen::data::JointStateRecord record;
  record.ts = trossen::data::make_timestamp_now();
  record.seq = 0;
  record.id = "demo/joints";
  record.positions = {0.1f, 0.2f, 0.3f};

  backend->write(record);
  backend->flush();
  backend->close();

  std::cout << "Successfully wrote data through registry-created backend!\n\n";

  return 0;
}

File: python/trossen_sdk/demo/backend_registry_demo.py
# Copyright 2025 Trossen Robotics
#
# Purpose:
# This script demonstrates the BackendRegistry system — creating backends at
# runtime without hardcoded type checking.

# The script does the following:
# 1. Lists registered backend types
# 2. Creates a NullBackend via the registry
# 3. Writes a JointStateRecord through the backend
# 4. Closes the backend

import trossen_sdk


def main():
    print("\nBackend Registry Demo")
    print("=====================\n")

    # 1. Check which backends are registered
    print("Checking registered backends:")
    for backend_type in ("null", "trossen_mcap", "lerobot_v2"):
        registered = trossen_sdk.BackendRegistry.is_registered(backend_type)
        if registered:
            print(f"  Is Registered: {backend_type}")
    print()

    # 2. Create a null backend via the registry
    backend = trossen_sdk.BackendRegistry.create("null")
    backend.open()

    # 3. Write some data
    record = trossen_sdk.JointStateRecord(
        ts=trossen_sdk.make_timestamp_now(),
        seq=0,
        id="demo/joints",
        positions=[0.1, 0.2, 0.3],
        velocities=[0.0, 0.0, 0.0],
        efforts=[0.0, 0.0, 0.0],
    )

    backend.write(record)
    backend.flush()
    backend.close()

    print("Successfully wrote data through registry-created backend!\n")


if __name__ == "__main__":
    main()
